workflow-bwa-mem/src/test/java/ca/on/oicr/pde/workflows/BwaMemWorkflowTest.java

```diff
@@ -49,7 +49,7 @@ public void dryRunRegressionTests() throws IOException, Exception {
             Map<String, String> params = new HashMap<>(t.getParameters());
             DryRun d = new DryRun(System.getProperty("bundleDirectory"), params, BwaMemWorkflow.class);
             d.buildWorkflowModel();
-            d.validateWorkflow();
+            //d.validateWorkflow();
```

Comment:
Insides `validateWorkflow()` there is a line `Assert.assertFalse(c.contains("null"), "Warning: command contains \"null\":\n" + c + "\n");` which is to check whether a command has "null" or not. I commented out this method because a command for calling cutadapt contains "/dev/null" which will cause this test to fail. I do not have access to this `d.validateWorkflow()` method, so if anyone could please change that assertion in the body of `d.validateWorkflow()`, I can uncomment it.

Reply:
There are a few other things we should try before we go modifying the methods in our standard DryRun library. For instance, would you run into any issues if you were to omit the `-o` parameter and send the output to standard out?

Reply:
If the `-o` parameter was omitted, then the output would be saved in the redirected file (i.e. in this case, the file of `logFilePath` on the line 259 will store all trimmed data instead of the statistics). If there is a place on the cluster to where the output can be saved, and the path does not contain "null", that will work.

Reply:
You might override the method in DryRun in an anonymous subclass... https://github.com/oicr-gsi/pipedev/blob/master/pipedev-test-utils/src/main/java/ca/on/oicr/pde/testing/workflow/DryRun.java

it's been a while, but something like this?

```
DryRun d = new DryRun(System.getProperty("bundleDirectory"), params, BwaMemWorkflow.class) {
  @Override
  public void validateWorkflow() {

    //check for null string
    for (AbstractJob j : w.getWorkflow().getJobs()) {

        String c = Joiner.on(" ").useForNull("null").join(j.getCommand().getArguments());
        # support piping output to /dev/null for cutadapt log files
        # use with caution!
        c = c.replace("/dev/null", "/tmp/junk");
        //check for null string
        Assert.assertFalse(c.contains("null"), "Warning: command contains \"null\":\n" + c + "\n");

        // check for missing spaces
        Assert.assertFalse(c.matches("(.*)[^ ]--(.*)"));
      }

      //view output files
      for (AbstractJob j : w.getWorkflow().getJobs()) {
        for (SqwFile f : j.getFiles()) {
          if (f.isOutput()) {
              System.out.println(f.getProvisionedPath());
          }
        }
      }
    }
};
```

Reply:
`w` is a private variable in the `DryRun` class (`private AbstractWorkflowDataModel w;`), so it failed to be overridden. Not sure if the access modifier of this variable can be changed to `protected` to make override possible.

Reply:
Whee, hacking Java... You can get a reference to `w` when you run `buildWorkflowModel`. How about something like this?

```
DryRun d = new DryRun(System.getProperty("bundleDirectory"), params, BwaMemWorkflow.class) {

  AbstractWorkflowDataModel wfModel;

  @Override
  public AbstractWorkflowDataModel buildWorkflowModel {
     wfModel = super.buildWorkflowModel();
     return wfModel;
  }

  @Override
  public void validateWorkflow() {

    //check for null string
    for (AbstractJob j : wfModel.getWorkflow().getJobs()) {

        String c = Joiner.on(" ").useForNull("null").join(j.getCommand().getArguments());
        # support piping output to /dev/null for cutadapt log files
        # use with caution!
        c = c.replace("/dev/null", "/tmp/junk");
        //check for null string
        Assert.assertFalse(c.contains("null"), "Warning: command contains \"null\":\n" + c + "\n");

        // check for missing spaces
        Assert.assertFalse(c.matches("(.*)[^ ]--(.*)"));
      }

      //view output files
      for (AbstractJob j : wfModel.getWorkflow().getJobs()) {
        for (SqwFile f : j.getFiles()) {
          if (f.isOutput()) {
              System.out.println(f.getProvisionedPath());
          }
        }
      }
    }
};
```

Reply:
Using a reference to `w` works fine. Thanks for the suggestion.